Tdk invensense icm42x70 support 3axis

afontaine-invn opened the pull request:
The icm42370-p is a 3-axis MEMS accelerometer.
https://invensense.tdk.com/products/motion-tracking/3-axis/icm-42370-p/
To support icm42370-p sensor, we renamed icm42670 to icm42x70 as the driver is similar to ICM-42670-P/S in hal_tdk module.
Add icm42x70 device tree file, it includes the common properties for icm42370-p, icm42670-p, and icm42670-s. Add icm42370-p device tree files for SPI and I2C interface.

Signed-off-by: Aurelie Fontaine [aurelie.fontaine@tdk.com](mailto:aurelie.fontaine@tdk.com)

kartben reviewed (dismissed):
Great work!
Haven't had a deep look from an actual driver perspective but my initial comment would be that you need to please describe in the Zephyr 4.1 migration guide the changes users of the existing driver/compatible will have to make to their codebase since things will "break" for them due to some of the refactoring -- it looks like it might be pretty minimal (so kudos on that!) but you should e.g. describe things like the fact that the public header for sensor custom attribute has been renamed include/zephyr/drivers/sensor/icm42670.h → include/zephyr/drivers/sensor/icm42x70.h ; things like that.

pdgendt commented on modules/hal_tdk/Kconfig:
```diff
@@ -17,4 +17,10 @@ config USE_EMD_ICM42670
 		|| DT_HAS_INVENSENSE_ICM42670S_ENABLED
 
+config USE_EMD_ICM42370
+	bool "ICM42370P High Performance 3-Axis MotionTracking Accelerometer"
```
Add `default y` similar to `USE_EMD_ICM42670`

afontaine-invn replied:
Thanks for pointing me this, this is updated

afontaine-invn commented:
Hi @MaureenHelm ,
Could you have a look at this pull request?
I added this change in release note document as requested and I've got conflicts to resolve many times lately that's why I needed to repsuh several times but nothing else changed.

yperess commented on drivers/sensor/tdk/icm42x70/icm42x70.c:
```diff
@@ -377,7 +381,8 @@ static int icm42670_accel_config(struct icm42670_data *drv_data, enum sensor_att
 }
 
-static int icm42670_set_gyro_odr(struct icm42670_data *drv_data, const struct sensor_value *val)
+#if CONFIG_USE_EMD_ICM42670
```
Suggestion, to make the code more readable add these functions to `icm42670.h` and include it from `icm42x70.h`, then move the definitions to an `icm42670.c` file and guard the inclusion of it in the cmake file. I generally find that it's easier to read the code that way.

yperess commented on drivers/sensor/tdk/icm42x70/icm42x70.c:
```diff
@@ -829,57 +854,60 @@ static int icm42670_fetch_from_registers(const struct device *dev, enum sensor_c
-	    (chan == SENSOR_CHAN_GYRO_X) || (chan == SENSOR_CHAN_GYRO_Y) ||
-	    (chan == SENSOR_CHAN_GYRO_Z) || (chan == SENSOR_CHAN_DIE_TEMP)) {
-#ifdef CONFIG_ICM42670_TRIGGER
-		status = icm42670_fetch_from_fifo(dev);
+	    (chan == SENSOR_CHAN_ACCEL_Z)
```
nit: prefer `sensor.h`'s `SENSOR_CHANNEL_IS_ACCEL(chan)` and `SENSOR_CHANNEL_IS_GYRO(chan)` for readability.

yperess reviewed (dismissed):
I have some minor nits about the way `#ifdef` is used where I think it would be much easier to read if we used `if (IS_ENABLED(...))` instead, but this PR is migrating existing code that's already been reviewed so I wouldn't require any cleanup.

MaureenHelm commented on drivers/sensor/tdk/icm42x70/icm42x70.c:
```diff
@@ -490,6 +397,7 @@ static int icm42670_turn_on_sensor(const struct device *dev)
 	}
 	LOG_DBG("Set accel full scale to: %d G", data->accel_fs);
 
+#if CONFIG_USE_EMD_ICM42670
```
This won't work if there is both an icm42670 and an icm42370 in the system. You need to check if *this* device is an icm42670, not if *any* device is an icm42670.

MaureenHelm replied:
You need to store the result of the DT_COMPAT test in the device config struct, then do a runtime check (not a preprocessor check) in the driver function.

afontaine-invn replied:
Ok, thank you for your comment. I kept both checks: a runtime check and a preprocessor check. This way it allows to build a lighter driver for icm42370 alone, and to call the correct APIs using icm42670 and icm42370 in a same system.

MaureenHelm replied:
Looks good now, thanks.

MaureenHelm reviewed (changes requested):
Most of the changes look fine, but the `#if CONFIG_USE_EMD_ICM42670` parts need to be reworked to check on a per-device basis.

yperess commented on drivers/sensor/tdk/icm42x70/icm42670.c:
```diff
@@ -0,0 +1,137 @@
+
+static int icm42670_set_gyro_odr(struct icm42x70_data *drv_data, const struct sensor_value *val)
+{
+	if (val->val1 <= 1600 && val->val1 > 12) {
```
Should this be `val->val1 >= 12`? Looking at the data sheet we can set 12.5Hz so I think 12 is OK.

yperess commented on drivers/sensor/tdk/icm42x70/icm42670.c:
```diff
@@ -0,0 +1,137 @@
+
+static int icm42670_set_gyro_fs(struct icm42x70_data *drv_data, const struct sensor_value *val)
+{
+	if (val->val1 > 2000 || val->val1 < 250) {
+		LOG_ERR("Incorrect fullscale value");
+		return -EINVAL;
```
`/** Sensor range, in SI units. */`

2000 and 250 are in degrees. You should expect the value in radians.

afontaine-invn replied:
It's a bug present in the current driver so thanks for this review. This is fixed now.

yperess reviewed (changes requested):
Looking pretty good, found an issue with the full scale units.

MaureenHelm reviewed (commented):
Need to address the twister failure and @yperess 's feedback, but otherwise looks good

afontaine-invn committed:
dts: bindings: sensor: add icm42370-p support

Add icm42x70 device tree file, it includes the common properties
for icm42370-p, icm42670-p, and icm42670-s.
Add icm42370-p device tree files for SPI and I2C interface.

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

afontaine-invn committed:
samples: sensor: 6dof_motion_drdy: update naming

Rename ICM42670 define to ICM42X70 to be aligned with updated driver
to support icm42370-p

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

afontaine-invn committed:
samples: sensor: tdk_apex: update naming

Rename ICM42670 define to ICM42X70 to be aligned with updated driver
to support icm42370-p

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

afontaine-invn committed:
tests: drivers: sensor: update naming

Rename ICM42670 define to ICM42X70 to be aligned with updated driver
to support icm42370-p

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

afontaine-invn commented:
> Need to address the twister failure and @yperess 's feedback, but otherwise looks good

Thanks, theses feedback and failure has been addressed.

afontaine-invn committed:
modules: hal_tdk: Add icm42370p support

The icm42370-p is a 3-axis MEMS accelermeter.
https://invensense.tdk.com/products/motion-tracking/3-axis/icm-42370-p/

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>

afontaine-invn committed:
drivers: sensor: icm42670: add icm42370-P support

To support icm42370-p sensor, we renamed icm42670 to icm42x70
 as the driver is similar to ICM-42670-P/S in hal_tdk module.
Keep icm42670 source and header for dedicated code.

Signed-off-by: Aurelie Fontaine <aurelie.fontaine@tdk.com>
